cns/restserver/ipam.go

```diff
@@ -118,6 +119,71 @@ func (service *HTTPRestService) GetPodIPConfigState() map[string]cns.IPConfigura
 	return service.PodIPConfigState
 }
 
+func (service *HTTPRestService) getIPAddressesHandler(w http.ResponseWriter, r *http.Request) {
+	var (
+		req           cns.GetIPAddressesRequest
+		resp          cns.GetIPAddressStateResponse
+		statusCode    int
+		returnMessage string
+		err           error
+	)
+
+	statusCode = UnexpectedError
+
+	defer func() {
+		if err != nil {
+			resp.Response.ReturnCode = statusCode
+			resp.Response.Message = returnMessage
+		}
+
+		err = service.Listener.Encode(w, &resp)
+		logger.Response(service.Name, resp, resp.Response.ReturnCode, ReturnCodeToString(resp.Response.ReturnCode), err)
+	}()
+
+	err = service.Listener.Decode(w, r, &req)
+	logger.Request(service.Name, &req, err)
+	if err != nil {
+		returnMessage = err.Error()
+		return
+	}
+
+	// Get all IPConfigs matching a state, and append to a slice of IPAddressState
+	resp.IPAddresses = make([]cns.IPAddressState, 0)
+	for _, podstate := range req.IPConfigStateFilter {
+		resp.IPAddresses = append(resp.IPAddresses, filterIPConfigsMatchingState(service.PodIPConfigState, podstate, func(ipconfig cns.IPConfigurationStatus, state string) bool {
+			return ipconfig.State == podstate
+		})...)
+	}
+
+	return
+}
+
+// filter the ipconfigs in CNS matching a state (Available, Allocated, etc.) and return in a slice
+func filterIPConfigsMatchingState(toBeAdded map[string]cns.IPConfigurationStatus, state string, f func(cns.IPConfigurationStatus, string) bool) []cns.IPAddressState {
```

Comment:
The name seems odd given you are passing a func with which you could do the opposite :) and it also transforms the input. There are a few times when I miss generics and working with collections in a functional manner is one of them. Maybe a predicate function is not needed if we always want to filter based on a match? Would clean up the above too.